Add test project

Stan-Kudri opened the pull request:
2 Коммита.

Stan-Kudri committed:
Добавил новый проект, с логикой из класса Calculator.

Добавил "{}" в else if.

Изменил зависимости проектов друг от друга.

Stan-Kudri committed:
Изменил кодировку на UTF-8.

blowin commented on TestProjectCalculator/UnitTest1.cs:
```diff
@@ -0,0 +1,15 @@
+{
+    public class UnitTest1
+    {
+        /*
```
Хорошо бы реализовать

blowin commented on Calculator.Domain/MathCalculator.cs:
```diff
@@ -78,19 +82,14 @@ private static decimal Eval(decimal a, decimal b, char operation)
                         {
-                            MessageBox.Show("Деление на 0 невозможно");
-                            return 0;
+                            throw new ArgumentException("Деление на 0 невозможно");
```
Можно бросать специальное исключение и в приложении отлавливать его и отображать сообщение

blowin reviewed (commented):
Доработай

Stan-Kudri committed:
Добавил пару тестов.

Изменил именование переменных.

Добавил проверку оператора, если создали класс и в первый раз обратились к методам, тогда прошлый оператор пуст, и в начале метода ему присваивается переданный оператор.

Stan-Kudri committed:
Изменил названия методов в MathCalculator.

Убрал коментарии с TryParse, предварительно изменив NumberStyles.

blowin commented on Calculator.Domain/MathCalculator.cs:
```diff
@@ -7,7 +9,7 @@ public class MathCalculator
         private decimal _valueOperand;
 
-        public decimal ArithmeticOperation(string outputValue, string inputValueOperand, char operation)
+        public decimal Eval(string outputValue, string inputValueOperand, char operation)
```
переименуй left и right, будет понятнее

blowin commented on Calculator.Domain/MathCalculator.cs:
```diff
@@ -16,10 +18,10 @@ public decimal ArithmeticOperation(string outputValue, string inputValueOperand,
                 throw new ArgumentException("Неверная арифметическая операция");
 
-            if (decimal.TryParse(outputValue /*, NumberStyles.Number, CultureInfo.InvariantCulture*/, out var firstValue))
+            if (decimal.TryParse(outputValue, NumberStyles.AllowDecimalPoint, CultureInfo.InvariantCulture, out var firstValue))
```
вынеси код по парсингу

blowin commented on Calculator.Domain/MathCalculator.cs:
```diff
@@ -78,7 +80,7 @@ public decimal GetNumericResult(string outputValue, string inputValueOperand)
         }
 
-        private static decimal Eval(decimal outputValue, decimal inputValueOperand, char operation)
+        private static decimal ArithmeticOperation(decimal outputValue, decimal inputValueOperand, char operation)
```
Calculate

blowin replied:
ArithmeticOperation - не несет смысла, у тебя метод, т.е действие. А сейчас у тебя метод "Арифметическая операция", логичнее было бы вызвать метод подсчёта

blowin commented on Calculator/FormCalculator.cs:
```diff
@@ -65,7 +65,7 @@ private void btnEqualSign_Click(object sender, EventArgs e)
             {
                 var tupleValuePad = ValueTxt();
```
Название ValueTxt очень плохое, вообще не понятно, что это такое

Stan-Kudri committed:
Добавил в форму кнопок арифметических операций и равно try-catch-finally.

Stan-Kudri committed:
Добавил тест, который проверяет правильность вычитания. #1

Stan-Kudri committed:
Удалил ненужную проверку _pastOperation, которую добавлял для реализации тестов.

Поле _pastOperation имеет дефолтное значение, при создании класса MathCalculator.

Добавил проверку в методы Evil на _pastOperation. Если _pastOperation не входит в массив операций, тогда вычисления не выполняются, а проходят дальше.

Test:
UnitTestOperation:
Добавил в каждый тест переменную expectResult, которая теперь не вычисляется в тесте, а имеет значение в атрибуте InlineData.

Добавил, грубо говоря, к каждой итерации +1 (т.е. заменил < на <=), для того, чтобы реализовался правильно метод, как при использовании калькулятора. (Действие происходит при нажатии следующей арифметической операции или равно)

UnitTestCustomCalculations: #1
Добавил новый тест, который включает 2 теста:
1 - Использование знака суммы, при имеющемся значении, но без операнда с последующим использованием знака '=' в количестве значения итераций;
1 - Использование знака суммы, при имеющемся значении и значением операнда с последующим использованием знака '=' в количестве значения итераций;

blowin commented on Calculator/FormCalculator.cs:
```diff
@@ -49,26 +49,58 @@ private void btnArithmeticOperation_Click(object sender, EventArgs e)
-                txtScoreboard.Text = string.Empty;
+                try
+                {
+                    var tupleValuePad = ValueTxt();
```
уже не в первый раз пишу, что название ValueTxt очень плохое, назови правильно

blowin commented on Calculator/FormCalculator.cs:
```diff
@@ -49,26 +49,58 @@ private void btnArithmeticOperation_Click(object sender, EventArgs e)
+                    var tupleValuePad = ValueTxt();
+                    var operation = char.Parse(((Button)sender).Text);
+
+                    var valueTxtBox = _calculator.Eval(tupleValuePad.textFromOutputField, tupleValuePad.textFromInputPad, operation);
```
используй сразу деконструкцию, чтобы не обращаться к полям через точку
https://docs.microsoft.com/en-us/dotnet/csharp/fundamentals/functional/deconstruct

blowin commented on Calculator/FormCalculator.cs:
```diff
@@ -49,26 +49,58 @@ private void btnArithmeticOperation_Click(object sender, EventArgs e)
+                    txtBoxValue.Text = string.Empty;
+                    MessageBox.Show(ex.Message);
+                }
+                catch (OperationCanceledException ex)
```
а если будет что-то другое, не OperationCanceledException? Отлавливай Exception

blowin commented on TestProjectCalculator/UnitTestCustomCalculations.cs:
```diff
@@ -0,0 +1,43 @@
+        [Theory]
+        [InlineData(21, 5, 126)]
+        [InlineData(0, 4, 0)]
+        public void Equal_To_Summation_Without_Operand_Value(decimal valueResult, int NumberIterations, decimal expectResult)
```
код стайл по именованию параметров

blowin commented on TestProjectCalculator/UnitTestCustomCalculations.cs:
```diff
@@ -0,0 +1,43 @@
+
+            valueResult = mathCalculator.Eval(valueResult.ToString(), string.Empty, '+');
+
+            for (var i = 0; i < NumberIterations; i++)
```
зачем?

blowin commented on TestProjectCalculator/UnitTestCustomCalculations.cs:
```diff
@@ -0,0 +1,43 @@
+        [Theory]
+        [InlineData(21, 4, 5, 41)]
+        [InlineData(0, 3, 4, 12)]
+        public void Equal_To_Summation_With_Operand_Value(decimal valueResult, decimal valueOperand, int NumberIterations, decimal expectResult)
```
Ты проверяешь работу с предыдущим оператором, так отрази это в названии

blowin commented on TestProjectCalculator/UnitTestCustomCalculations.cs:
```diff
@@ -0,0 +1,43 @@
+            valueResult = mathCalculator.Eval(valueResult.ToString(), valueOperand.ToString(), '+');
+            valueResult = mathCalculator.Eval(valueResult.ToString(), valueOperand.ToString());
+
+            for (var i = 0; i < NumberIterations - 1; i++)
```
Вопрос тот же

Stan-Kudri committed:
Изменил название Кортежа и его состовляющих.

Использовал деконструкцию кортежей.

Stan-Kudri committed:
Изменил тип исключения при делении на ноль и заменил тип второго отлавливаемого исключения на Exception.

Stan-Kudri committed:
Изменил названия методов и переменных(параметров).

blowin commented on Calculator.Domain/MathCalculator.cs:
```diff
@@ -73,33 +73,33 @@ public decimal Eval(string outputValue, string inputValueOperand)
                 case '*':
                     {
-                        return outputValue == 0 ? outputValue : outputValue * inputValueOperand;
+                        return left == 0 ? left : left * right;
```
зачем эта проверка? умножай просто и так будет 0

blowin commented on Calculator/FormCalculator.cs:
```diff
@@ -116,7 +116,7 @@ private void txtScoreboard_KeyPress(object sender, KeyPressEventArgs e)
         }
 
-        private (string textFromInputPad, string textFromOutputField) ValueTxt() => (txtScoreboard.Text, txtBoxValue.Text);
+        private (string textFromInput, string textFromOutput) Scoreboard() => (txtScoreboard.Text, txtBoxValue.Text);
```
дай лучше название. какое-нибудь, GetValuesToCount
и названия можешь сразу дать left и right

blowin commented on Calculator/FormCalculator.cs:
```diff
@@ -81,18 +81,18 @@ private void btnEqualSign_Click(object sender, EventArgs e)
                     txtBoxValue.Text = valueTxtBox.ToString();
                 }
-                catch (ArgumentException ex)
+                catch (DivideByZeroException ex)
```
Бросай какое-нибудь кастомное, одно и то же, и показывай его сообщение, а если я скажу добавить другую проверку и ты бросишь ещё одно исключение и по факту тебе его тоже нужно отобразить, будешь писать ещё 1 catch?
https://docs.microsoft.com/en-us/dotnet/api/system.applicationexception?view=net-6.0

Stan-Kudri committed:
Изменил названия переменных кортежа и именование Кортежа.

Добавил приватный метод TryParseValue, который выполняет прошлую проверку.

Stan-Kudri committed:
Добавил класс ApplicationExceprion, который выпадает при делении числа на ноль.

Stan-Kudri committed:
Добавил тесты #1 для проверки операций: "-", "*", "/".

Stan-Kudri committed:
Добавил тест, который проверяет выподающее исключение, при делении на ноль.

Stan-Kudri committed:
Добавил тест, в котором выпадает исключение, если в табле находятся не чисал.

Изменил в классе MathCalculator условия выподения ошибки:
ArgumentException("Неправильная числовая строка").

Добавил условие при реализации "=", если прошлого оператора небыло, а точнее было только =;

blowin commented on Calculator.Domain/ApplicationException.cs:
```diff
@@ -0,0 +1,17 @@
+﻿namespace Calculator.Domain
+{
+    public class ApplicationException : Exception
```
Почему не использовать готовый?

blowin commented on TestProjectCalculator/UnitTestByExceptionMathCalculator.cs:
```diff
@@ -0,0 +1,36 @@
+
+            valueResult = mathCalculator.Eval(valueResult.ToString(), "0", '/');
+
+            Assert.Throws(expectType, () =>
```
Можно указать какой тип исключения получишь

blowin commented on TestProjectCalculator/UnitTestCustomCalculations.cs:
```diff
@@ -39,5 +39,108 @@ public void Performing_Summation_Without_Operand_When_Pressing_Equals(decimal va
+
+            for (var i = 0; i < numberIterations - 1; i++)
+            {
+                valueResult = mathCalculator.Eval(valueResult.ToString(), string.Empty);
```
Ты так и не объяснил зачем циклы

blowin commented on Calculator/FormCalculator.cs:
```diff
@@ -81,13 +82,13 @@ private void btnEqualSign_Click(object sender, EventArgs e)
                 try
                 {
-                    var (textFromInput, textFromOutput) = Scoreboard();
+                    var (textFromInput, textFromOutput) = GetValuesToCount();
```
при деконструкции тоже давай нормальные имена во всех местах использования

blowin commented on TestProjectCalculator/UnitTestByExceptionMathCalculator.cs:
```diff
@@ -0,0 +1,36 @@
+        {
+            var mathCalculator = new MathCalculator();
+
+            valueResult = mathCalculator.Eval(valueResult.ToString(), "0", '/');
```
при вызове ToString(ВЕЗДЕ) нужно использовать правильную культуру
https://docs.microsoft.com/en-us/dotnet/api/system.globalization.cultureinfo.invariantculture?view=net-6.0

blowin reviewed (commented):
Ты регулярно пропускаешь некоторые комментарии. Будь внимательнее

Stan-Kudri committed:
Удалил свой ApplicationException.

Stan-Kudri committed:
Добавил CultureInfo.InvariantCulture в ToString().

Дал имена переменым кортежей, там где он создается.

blowin commented on Calculator/FormCalculator.cs:
```diff
@@ -82,9 +81,9 @@ private void btnEqualSign_Click(object sender, EventArgs e)
+                    var (left, right) = GetValuesToCount();
 
-                    var valueTxtBox = _calculator.Eval(textFromOutput, textFromInput);
+                    var valueTxtBox = _calculator.Eval(right, left);
```
Переименуй и параметры Eval и передавай туда left и right

Stan-Kudri committed:
Изменения.

Добавил ещё один флаг NumberStyles в TryParse, для учета знака.

Добавил в тесты приватный метод, который в входных параметрах получает число типа decimal, а на выходе строку c свойством CultureInfo.InvariantCulture.
